STCOM-534: Card component

doytch opened the pull request:
- Added new `Card` component to facilitate reuse of styles when creating card-based layouts including those [with remote-resource-lookup patterns](https://issues.folio.org/browse/UXPROD-1681).
- Has `default`, `positive` and `negative` styles to apply background styles.
- Has `roundedBorder` and `hasMargin` props to compose styles.
- Supports application of classes directly onto the card, header, and body elements.
- Complete rendering control of header and body.

<img width="656" alt="cards" src="https://user-images.githubusercontent.com/5324374/58501980-9266f380-8153-11e9-8634-308adeef7bdc.png">

id-jenkins commented:
<html><body><pre>
yarn run v1.9.4
$ eslint ./ && stylelint "lib/**/*.css"

lib/Card/Card.css
 71:1  ✖  Unexpected missing end-of-source newline   no-missing-end-of-source-newline

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>

JohnC-80 commented on lib/Card/readme.md:
```diff
@@ -0,0 +1,36 @@
+
+## `cardStyle`
+
+The `default` card style is selected automatically. Optionally, you can set the `cardStyle` prop to `lookupPreview` and `lookupEmpty` for easy application of the styles required for a remote resource lookup pattern.
```
Is there a more defined system-standard for these styles where we could say "use 'lookupPreview' style for this state, Use 'lookupEmpty' ... etc"
Maybe even an example of code that would apply the prop based on the presence of a resource. @filipjakobsen ?

doytch replied:
Yeah I'd love to know the specific nomenclature we'll be using for this going forward as well since I don't know it :) 

I don't know exactly how we'd automatically apply it though. The place where we'd use it would be [something like here](https://github.com/folio-org/ui-agreements/blob/master/src/components/Agreements/ViewAgreement/Sections/Organizations.js#L40).

JohnC-80 replied:
hmm... without particular meaning for these specific names, I'd probably go with something more generic `bordered` and `solid` perhaps since the prop itself just assigns a preset style... With that, I'd say we could merge this in.

JohnC-80 commented on lib/Card/Card.js:
```diff
@@ -0,0 +1,105 @@
+          data-test-card-body
+          className={this.getBodyStyle()}
+        >
+          {this.props.children}
```
Minor, but maybe destructure `headerEnd`,`headerStart`, and `children`

filipjakobsen commented:
I actually don't think the empty state should be a part of the card itself, but rather be a part of whichever area the card exists in. In some places, users will have to add multiple cards in one section. This parent section may then be empty or not.

filipjakobsen commented:
If the two different styles are intended to support nesting, then "figure" and "ground" work, respectively; or "positive" and "negative". The "bordered" version is actually not just outlined; it also has a white background, which will be visible when place on top of a grey area.

rasmuswoelk commented on lib/Card/Card.css:
```diff
@@ -0,0 +1,71 @@
+  border-width: 1px;
+  margin-bottom: var(--gutter);
+
+  &.default {
```
I think we could avoid nesting here to make the CSS less specific (making it easier for devs to overwrite styles if needed).

This class will end up being `.card.default` – instead, we could just target `.default`.

The goes for other nested style class names.

rasmuswoelk commented on lib/Card/Card.css:
```diff
@@ -0,0 +1,71 @@
+    }
+  }
+
+  &.lookupEmpty,
```
I would definitely prefer some more generic style names – e.g. "bordered" and "radius" or similar.

rasmuswoelk commented on lib/Card/Card.css:
```diff
@@ -0,0 +1,71 @@
+  display: flex;
+  justify-content: flex-start;
+  align-items: center;
+  padding: 8px var(--gutter);
```
Here I would recommend using the `--gutter-static` variable instead of `--static`. The static variable doesn't expand if the root font-size changes.

Instead of the 8px vertical padding, you can use `--gutter-static-one-third`. This would add 5px instead of 8px. If you want the header height to be a little bit bigger then you could add a min-height. Smaller vertical padding combined with a min-height would also ensure that the header doesn't get too tall if the content inside is big (making the distance between the body and the header too big).

rasmuswoelk commented on lib/Card/Card.js:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  static defaultProps = {
+    cardStyle: 'default',
```
If the default style is "default" and this style adds margin-bottom: 0; then the `marginBottom0`-class name becomes obsolete unless you actually add a`cardStyle`, right? 

In that case, maybe the card shouldn't have a "cardStyle" default prop? Coming back to the more generic style names then we could maybe have a "nested" style that removes the bottom-margin for all nested cards except the last one. And maybe also removes the top-border for all .nested cards except the first one.

rasmuswoelk commented on lib/Card/Card.css:
```diff
@@ -0,0 +1,71 @@
+    border-width: 1px 1px 0;
+    margin-bottom: 0;
+
+    &:last-of-type {
```
The "default" style card has no border-bottom which looks odd if there's only one rendered card.

Could we instead do something like:

```css
  .default + .default {
     border-top: 0;
   }
```

This way a single "default"-styled card will have a border-bottom and when multiple cards are nested it removes the border-top instead to avoid double border.

id-jenkins commented:
<html><body><pre>
yarn run v1.9.4
$ eslint ./ && stylelint "lib/**/*.css"

/home/jenkins/workspace/io-org_stripes-components_PR-976/project/lib/Card/DefaultCardHeader.js
   5:8  warning  'omitProps' is defined but never used          no-unused-vars
  58:2  error    Newline required at end of file but not found  eol-last

✖ 2 problems (1 error, 1 warning)
  1 error and 0 warnings potentially fixable with the `--fix` option.

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>

id-jenkins commented:
<html><body><pre>
yarn run v1.9.4
$ eslint ./ && stylelint "lib/**/*.css"

/home/jenkins/workspace/io-org_stripes-components_PR-976/project/lib/Card/DefaultCardHeader.js
   5:8  warning  'omitProps' is defined but never used          no-unused-vars
  58:2  error    Newline required at end of file but not found  eol-last

✖ 2 problems (1 error, 1 warning)
  1 error and 0 warnings potentially fixable with the `--fix` option.

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>

id-jenkins commented:
<html><body><pre>
yarn run v1.9.4
$ eslint ./ && stylelint "lib/**/*.css"

lib/Card/DefaultCardHeader.css
 18:1  ✖  Unexpected missing end-of-source newline   no-missing-end-of-source-newline

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>

doytch commented:
Refactored based on the suggestions. Main things are:

- Removed the concept of `lookupPreview` and `lookupEmpty` for the cards. Those styles can now be achieved via setting a `cardStyle` of `positive` or `negative` and applying `hasMargin` and `roundedBorder` props. While I originally wanted to be able to turn on all the styles with one prop, I get why more discrete props can be cleaner.
- I ensured that the default style that's most likely to be used (the repeating card pattern) is the one that requires the fewest props.
- Split out a `DefaultCardHeader` component.

rasmuswoelk commented on lib/Card/Card.js:
```diff
@@ -0,0 +1,89 @@
+        <HeaderComponent {...this.props} />
+        <div
+          data-test-card-body
+          className={`${css.body} ${bodyClass || ''}`}
```
Maybe we can use the "classnames" package here instead?

doytch replied:
I thought about it but it seemed like overkill given that I'm literally just appending the one class.

rasmuswoelk replied:
Fair :) And you are probably right. I just like `classnames` because it automatically prevents undefined or empty classes to be appended.

rasmuswoelk commented on lib/Card/DefaultCardHeader.js:
```diff
@@ -0,0 +1,57 @@
+
+import css from './DefaultCardHeader.css';
+
+export default class DefaultCardHeader extends React.Component {
```
To keep the component folder clean; could we move this component into its own folder and then place it inside a "headers"-folder?

rasmuswoelk commented on lib/Card/readme.md:
```diff
@@ -0,0 +1,35 @@
+--- | --- | --- | --- |
+| `bodyClass` | string | |
+| `cardClass` | string | |
+| `cardStyle` | string | | Must be undefined or one of `default`, `negative`, or `positive`. These generally affect the background of the card.
```
Should we maybe elaborate these styles a bit in the readme?

rasmuswoelk commented on lib/Card/Card.js:
```diff
@@ -0,0 +1,89 @@
+        className={this.getCardStyle()}
+        {...customProps}
+      >
+        <HeaderComponent {...this.props} />
```
I'm wondering if we should only pass the props related to the header into the header component..  🤔What do you think?

doytch replied:
Yeah, but if we want to allow users to add arbitrary props to the header (`data-test-etc`) what are the props that are intended for the header? 

Sometimes people add a `headerProps` object to destructure but I'm not a huge fan of it because it grows the props list and it can get you into situations where you have to define a prop in `props` and `headerProps` if you want it going to the card and the header.

I don't feel too strongly about it here, but that's why I wrote it this way instead of the `headerProps` way after thinking about it for a bit. I'm willing to change it to something else though.

JohnC-80 commented:
Whatever we decide on for style names, the usage must have an explanation in the readme if we want the standard to be adopted in the system - they don't even need to be 1-to-1 use-cases, just need to be enough to help a developer (or PO) decide which they need - possibly even based on how @doytch 's module will use them - just with generic names that don't couple it up with the particular use-case. 
A `renderHeader` function that's passed the props might be a better prop option for max composability/separation of props for the header rather than leaving it only with the props passed to the Card.

doytch commented:
Added a **Common Patterns** section per @JohnC-80's suggestion
